rotors_gazebo_plugins/src/gazebo_wind_plugin.cpp

```diff
@@ -99,38 +169,160 @@ void GazeboWindPlugin::OnUpdate(const common::UpdateInfo& _info) {
-  if (now >= wind_gust_start_ && now < wind_gust_end_) {
-    double wind_gust_strength = wind_gust_force_mean_ + wind_gust_force_n_(random_generator_);
-    wind_gust = wind_gust_strength * wind_gust_direction_;
-    // Apply a force from the wind gust to the link.
-    link_->AddForceAtRelativePosition(wind_gust, xyz_offset_);
+  math::Vector3 wind_velocity = math::Vector3(0,0,0);
+
+  if (!custom_static_wind_field_) // Choose method
+  {
+    // Calculate the wind force.
+    double wind_strength = wind_force_mean_ + wind_force_n_(random_generator_);
+    math::Vector3 wind = wind_strength * wind_direction_;
+    // Apply a force from the constant wind to the link.
+    link_->AddForceAtRelativePosition(wind, xyz_offset_);
+
+    math::Vector3 wind_gust(0, 0, 0);
+    // Calculate the wind gust force.
+    if (now >= wind_gust_start_ && now < wind_gust_end_) {
+      double wind_gust_strength = wind_gust_force_mean_ + wind_gust_force_n_(random_generator_);
+      wind_gust = wind_gust_strength * wind_gust_direction_;
+      // Apply a force from the wind gust to the link.
+      link_->AddForceAtRelativePosition(wind_gust, xyz_offset_);
+    }
+
+    geometry_msgs::WrenchStamped wrench_msg;
+
+    wrench_msg.header.frame_id = frame_id_;
+    wrench_msg.header.stamp.sec = now.sec;
+    wrench_msg.header.stamp.nsec = now.nsec;
+    wrench_msg.wrench.force.x = wind.x + wind_gust.x;
+    wrench_msg.wrench.force.y = wind.y + wind_gust.y;
+    wrench_msg.wrench.force.z = wind.z + wind_gust.z;
+    wrench_msg.wrench.torque.x = 0;
+    wrench_msg.wrench.torque.y = 0;
+    wrench_msg.wrench.torque.z = 0;
+
+    wind_pub_.publish(wrench_msg);
+
+    // Calculate the wind speed
+    double wind_speed = wind_speed_mean_; // + wind_speed_n_(random_generator_);
+    wind_velocity = wind_speed * wind_direction_;
   }
+  else
+  {
+    // Get the current position of the aircraft in world coordinates
+    math::Vector3 link_position = link_->GetWorldPose().pos;
+
+    //Determine the wind velocity at the aircraft position if within world bounds. Else, set to user defined constant value.
+    if (!(link_position.x < coords_x_[0] || link_position.x > coords_x_[coords_x_.size()-1] || link_position.y < coords_y_[0] || link_position.y > coords_y_[coords_y_.size()-1] || link_position.z < z_min_ || link_position.z > z_max_))
+    {
+      // Create a vector containing the coordinates of 8 points corresponding to the vertices of the volume enclosing the link and the 5 points for interpolation
+      // Containing as well the wind velocity value for each point.
+      std::vector<math::Vector3> wind_field_cells[2];
+      for (int i = 0; i < 14; i++)
+        for (int j=0; j<2; j++)
+          wind_field_cells[j].push_back(math::Vector3(0, 0, 0));
+
+      // Find the vertices of the 8 points
+        // Find their x-coordinates
+      for (int i = 0; i < coords_x_.size(); i++)
+        if (coords_x_[i] < link_position.x && coords_x_[i+1] > link_position.x)
```

Comment:
I am confused about what grid choice you have made.

IF you assume an equidistant grid in x- and y-direction or more specifically an equally-spaced rectangular grid, then you should not even need to loop over all coords_x (which is extremely slow!) because your coords_x were already sorted in L.144, you know the min and max value and the number of elements, so you can directly _tell_ which index the link_position is at without looping.

However, the current readme.md does not give any restrictions on where the x,y,z points need to sit. However, if there is just "randomly spaced" x,y,z points in the wind_field_path file, then the assignment into one coord_y and one coord_y vector (see https://github.com/pvechersky/rotors_simulator/pull/5/files#diff-41259901fef6bb65e3af337b702a2c58R127) does not work anymore (because your x-locations are not the same over all y-locations).

Do you see what I mean? Or shall we discuss this during the next meeting offline?

It's maybe best if you add a proper description (maybe with a tiny drawing) of how the wind field position grid needs to look into the readme before I continue reviewing, because the remaining code of course heavily depends on that.